src/lib/coda_net2/coda_net2.ml

```diff
@@ -0,0 +1,1017 @@
+            [path ^/ libp2p_helper_binary]
+        | None ->
+            [] ) )
+        ~f:(fun prog -> Process.create ~prog ~args:[] ())
+      |> Deferred.Or_error.map ~f:(fun p ->
+             Helper.create logger p conf_dir lock_path )
+    with
+    | Ok p ->
+        (* If the libp2p_helper process dies, kill the parent daemon process. Fix
+       * for #550 *)
+        Deferred.upon (Process.wait p.subprocess) (fun code ->
+            p.finished <- true ;
+            ( match (p.failure_response, code) with
+            | `Ignore, _ | _, Ok () ->
+                ()
+            | `Die, (Error _ as e) ->
+                Logger.fatal logger ~module_:__MODULE__ ~location:__LOC__
+                  !"libp2p_helper process died: %s"
+                  (Unix.Exit_or_signal.to_string_hum e) ;
+                raise Child_died ) ;
+            Hashtbl.iter p.outstanding_requests ~f:(fun iv ->
+                Ivar.fill iv
+                  (Or_error.error_string
+                     "libp2p_helper process died before answering") ) ) ;
+        Ok p
+    | Error e ->
+        Or_error.error_string
+          ( "If you are a dev, did you forget to `make libp2p_helper` and set \
+             CODA_LIBP2P_HELPER_PATH? Try \
+             CODA_LIBP2P_HELPER_PATH=$PWD/src/app/libp2p_helper/result/bin/libp2p_helper "
+          ^ Error.to_string_hum e )
+  in
+  let kill_locked_process ~logger =
+    match%bind Sys.file_exists lock_path with
+    | `Yes -> (
+        let%bind p = Reader.file_contents lock_path in
+        match%bind Process.run ~prog:"kill" ~args:[p] () with
+        | Ok _ ->
+            Logger.debug logger ~module_:__MODULE__ ~location:__LOC__
+              "Killing dead libp2p_helper process %s" p ;
+            let%map () = Sys.remove lock_path in
+            Ok ()
+        | Error _ ->
+            Logger.debug logger ~module_:__MODULE__ ~location:__LOC__
+              "Process %s does not exist and will not be killed" p ;
+            return @@ Ok () )
+    | _ ->
+        return @@ Ok ()
+  in
+  let open Deferred.Or_error.Let_syntax in
+  let%bind () = kill_locked_process ~logger in
+  match%bind Sys.is_directory conf_dir |> Deferred.map ~f:Or_error.return with
+  | `Yes ->
+      let%bind t = run_p2p () in
+      let%map () =
+        write_lock_file lock_path (Process.pid t.subprocess)
+        |> Deferred.map ~f:Or_error.return
+      in
+      t
+  | _ ->
+      Deferred.Or_error.errorf "Config directory (%s) must exist" conf_dir
+
+let%test_module "coda network tests" =
+  ( module struct
+    let () = Backtrace.elide := false
+
+    let () = Async.Scheduler.set_record_backtraces true
+
+    let logger = Logger.create ()
+
+    let testmsg =
+      "This is a test. This is a test of the Outdoor Warning System. This is \
+       only a test."
+
+    let setup_two_nodes network_id =
+      let%bind a_tmp = Unix.mkdtemp "p2p_helper_test_a" in
+      let%bind b_tmp = Async.Unix.mkdtemp "p2p_helper_test_b" in
```

Comment:
why are you using two different `mkdtemp`s?

Reply:
This doesn't work on macos for some reason